### PR TITLE
[ASTWalker] Continue post-walking with `Action::SkipChildren`

### PR DESCRIPTION
**Pull request by hamishknight**
Previously we would skip the post-walk, which was not a particularly intuitive behavior. Introduce an `Action::SkipNode` variant for walkers that want to also skip the post-walk. This then lets up clean up some walkers that previously had to work around the old behavior by doing the post-walk manually.

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please test

File: lib/Sema/CSGen.cpp [hunk 1]
```diff
@@ -848,11 +844,7 @@ namespace {
...
     }
-    
-    PostWalkResult<Expr *> walkToExprPost(Expr *expr) override {
-      return Action::Continue(expr);
-    }
-    
+
     /// Ignore statements.
     PreWalkResult<Stmt *> walkToStmtPre(Stmt *stmt) override {
       return Action::SkipChildren(stmt);
```
Comment (xedin) [inline lib/Sema/CSGen.cpp:850]: I think both Stmt and Decl walkers here should use `SkipChildrenAndPostWalk`.

File: lib/Sema/CSGen.cpp [hunk 2]
```diff
@@ -124,11 +124,7 @@ namespace {
       
       return Action::Continue(expr);
     }
-    
-    PostWalkResult<Expr *> walkToExprPost(Expr *expr) override {
-      return Action::Continue(expr);
-    }
-    
+
```
Comment (xedin) [inline lib/Sema/CSGen.cpp:127]: All of the SkipChildren here should be `SkipChildrenAndPostWalk`, I don't think we care about post walk for any of the ignored things...

**Review by xedin — APPROVED** [review]

**Comment by hamishknight** [comment]
After talking with Pavel, I'm considering renaming `Action::SkipChildrenAndPostWalk` to something like `Action::SkipNode`, and making it the norm for ASTWalkers that don't care about the post-walk

**Review by bnbarham — APPROVED** [review]
Very nice 😍 

> I'm considering renaming Action::SkipChildrenAndPostWalk to something like Action::SkipNode, and making it the norm for ASTWalkers that don't care about the post-walk

This sounds nice to me.

**Comment by hamishknight** [comment]
> After talking with Pavel, I'm considering renaming Action::SkipChildrenAndPostWalk to something like Action::SkipNode, and making it the norm for ASTWalkers that don't care about the post-walk

Gone ahead and made this change

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
Looks great to me, thank you!

**Comment by hamishknight** [comment]
Going to hold off merging this for a little bit until we've stopped cherry-picking stuff to 5.10

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
🚢

**Pull request merged**